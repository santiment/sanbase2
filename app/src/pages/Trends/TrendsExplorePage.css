.TrendsExplorePage {
  margin: 30px 10%;
}

.TrendsExplorePage__content {
  background-color: #fff;
}

.TrendsExplorePage ul {
  list-style: none;
}
