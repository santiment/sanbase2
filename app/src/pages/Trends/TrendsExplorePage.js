import React, { Component } from 'react'
import PropTypes from 'prop-types'
import * as qs from 'query-string'
import TrendsExploreChart from '../../components/Trends/Explore/TrendsExploreChart'
import TimeFilter from './../../components/TimeFilter/TimeFilter'
import TrendsExploreHeader from '../../components/Trends/Explore/TrendsExploreHeader'
import TrendsExploreSourcesFilter from '../../components/Trends/Explore/TrendsExploreSourcesFilter'
import {
  validateSearchSources,
  Source
} from '../../components/Trends/trendsUtils'
import GetTrends from './../../components/Trends/GetTrends'
import './TrendsExplorePage.css'

export class TrendsExplorePage extends Component {
  state = {
    ...getStateFromQS(this.props)
  }

  static defaultProps = {
    match: { params: {} },
    location: {},
    history: {}
  }

  static propTypes = {
    match: PropTypes.object,
    location: PropTypes.object,
    history: PropTypes.object
  }

  static getDerivedStateFromProps (nextProps, prevState) {
    return {
      ...getStateFromQS(nextProps)
    }
  }

  render () {
    const { match, isDesktop } = this.props
    const { selectedSources } = this.state
    return (
      <div className='TrendsExplorePage'>
        <div className='TrendsExplorePage__content'>
          <TrendsExploreHeader
            topic={match.params.topic}
            selectedSources={selectedSources}
          />
          <TimeFilter
            timeOptions={['1w', '1m', '3m', '6m', '1y', 'all']}
            defaultSelected='6m'
            disabled
          />
          <GetTrends
            topic={match.params.topic}
            render={props => (
              <TrendsExploreChart
                {...props}
                isDesktop={isDesktop}
                selectedSources={selectedSources}
              />
            )}
          />
          <TrendsExploreSourcesFilter
            selectedSources={selectedSources}
            handleSourceSelect={this.handleSourceSelect}
          />
        </div>
      </div>
    )
  }

  handleSourceSelect = ({ currentTarget }) => {
    const { selectedSources } = this.state
    const source = currentTarget.dataset.source
    const newSelectedSource = calculateNewSources({
      source,
      selectedSources,
      sources: Object.keys(Source)
    })

    this.setState(
      {
        selectedSources: newSelectedSource
      },
      this.updateSearchQuery
    )
  }

  mapStateToQS = ({ interval, selectedSources }) =>
    '?' +
    qs.stringify(
      { interval, source: selectedSources },
      { arrayFormat: 'bracket' }
    )

  updateSearchQuery = () => {
    this.props.history.push({
      search: this.mapStateToQS(this.state)
    })
  }
}

export const getStateFromQS = ({ location }) => {
  const { interval, source } = qs.parse(location.search, {
    arrayFormat: 'bracket'
  })

  return {
    interval: interval || '6m',
    selectedSources: validateSearchSources(source)
  }
}

export const calculateNewSources = ({
  source,
  selectedSources = ['merged'],
  sources
}) => {
  if (source === 'merged') {
    if (selectedSources.includes('merged')) {
      return sources.filter(selectedSource => selectedSource !== 'merged')
    } else {
      return ['merged']
    }
  } else {
    if (selectedSources.includes(source)) {
      if (selectedSources.length === 1) {
        return ['merged']
      }
      return selectedSources.filter(selectedSource => selectedSource !== source)
    } else {
      return [...selectedSources, source]
    }
  }
}

export default TrendsExplorePage
