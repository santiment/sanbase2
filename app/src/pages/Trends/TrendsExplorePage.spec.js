/* eslint-env jest */
import React from 'react'
import toJson from 'enzyme-to-json'
import { shallow, mount } from 'enzyme'
import TrendsExplorePage from './TrendsExplorePage'

describe('TrendsExplorePage', () => {
  it('smoke', () => {
    const mockCb = jest.fn()
    const wrapper = shallow(<TrendsExplorePage />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  describe('calculateNewSources', () => {
    const sources = ['merged', 'telegram', 'reddit']

    it("it should return rest sources if clicked on 'merged' and 'merged' is actived", () => {
      const newSelected = TrendsExplorePage.calculateNewSources({
        sources,
        source: 'merged',
        selectedSources: 'merged'
      })
      expect(JSON.stringify(newSelected)).toBe(
        JSON.stringify(['telegram', 'reddit'])
      )
    })

    it("it should return merged sources if clicked on 'merged' and 'merged' is unactived", () => {
      const newSelected = TrendsExplorePage.calculateNewSources({
        sources,
        source: 'merged',
        selectedSources: ['reddit']
      })
      expect(JSON.stringify(newSelected)).toBe(JSON.stringify(['merged']))
    })

    it('it should return merged sources if clicked on any source and this source is last and single activated', () => {
      const newSelected = TrendsExplorePage.calculateNewSources({
        sources,
        source: 'reddit',
        selectedSources: ['reddit']
      })
      expect(JSON.stringify(newSelected)).toBe(JSON.stringify(['merged']))
    })
  })
})
