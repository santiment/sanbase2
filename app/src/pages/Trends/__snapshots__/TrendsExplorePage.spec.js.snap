// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrendsExplorePage smoke 1`] = `
<div
  className="TrendsExplorePage"
>
  <div
    className="TrendsExplorePage__content"
  >
    <TrendsExploreHeader
      selectedSources={
        Array [
          "merged",
        ]
      }
    />
    <TimeFilter
      defaultSelected="6m"
      disabled={true}
      timeOptions={
        Array [
          "1w",
          "1m",
          "3m",
          "6m",
          "1y",
          "all",
        ]
      }
    />
    <Apollo(Apollo(Apollo(GetTrends)))
      render={[Function]}
      selectedSources={
        Array [
          "merged",
        ]
      }
    />
    <TrendsExploreSourcesFilter
      handleSourceSelect={[Function]}
      selectedSources={
        Array [
          "merged",
        ]
      }
    />
  </div>
</div>
`;
